### PR TITLE
nixos/systemd: Support setting machineId

### PR DESCRIPTION
**Pull request by nicoonoclaste**
## Description of changes

- Add new `systemd.machineId` configuration option.
  In principle, [`machine-id(5)`] can be an arbitrary, 128b string in hex-encoding; however, systemd v30 onwards (2011) generate UUIDv4 values, and various tools may rely on that.

- Set the machine-id through `boot.kernelParams`.

[`machine-id(5)`]: https://www.freedesktop.org/software/systemd/man/latest/machine-id.html

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
- [ ] Tested, as applicable:
  - Didn't write a NixOS test for this  😅
- [x] Tested basic functionality
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Module updates) Added a release notes entry if the change is significant
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by aanderse** [comment]
i'm trying to understand the implications of this statement from the documentation you linked:

> This ID uniquely identifies the host. It should be considered "confidential", and must not be exposed in untrusted environments, in particular on the network.

do you know the details on that?

**Comment by flokli** [comment]
Quoting more from the manpage:

> The machine ID is usually generated from a random source during system installation or first boot and stays constant for all subsequent boots. Optionally, for stateless systems, it is generated during runtime during early boot if necessary.

We should make sure systemd is able to write to this during installation / first boot, but putting it into NixOS configuration sounds wrong.

For the special usecases where it's desired to set this via kernel cmdline, we already have the cmdline config option, but I wouldn't encourage further to set this.

**Comment by flokli** [comment]
> I set it through cmdline in the first place was so distinct machines with otherwise-identical config share more drvs (which is somewhat helpful in network-boot contexts etc.)

In the netboot scenario, can't you set / extend the kernel cmdline independently of the system closure? 

That's what I used to do, in cases where I wanted to pass along the ipxe machine UUID to systemd. Similar things were also suggested to people asking for this to be added into CoreOS:
https://github.com/coreos/bugs/issues/2166

> Are “impermanent” / root-on-tmpfs systems such a special usecase?

If you don't want to have a new machine UUID on such systems, simply exclude this file from being wiped on every reboot.

Having a static value part of your system config is only a big footgun, having two machines with the same UUID. I don't think we should introduce this as a config option.

File: nixos/modules/system/boot/systemd.nix [hunk 1]
```diff
@@ -629,7 +639,10 @@ in
     # https://github.com/systemd/systemd/pull/12226
     boot.kernel.sysctl."kernel.pid_max" = mkIf pkgs.stdenv.is64bit (lib.mkDefault 4194304);
 
-    boot.kernelParams = optional (!cfg.enableUnifiedCgroupHierarchy) "systemd.unified_cgroup_hierarchy=0";
+    boot.kernelParams = lib.foldr (a: b: a ++ b) [] [
+      (optional (cfg.machineId != null) "systemd.machine_id=${cfg.machineId}")
+      (optional (!cfg.enableUnifiedCgroupHierarchy) "systemd.unified_cgroup_hierarchy=0")
+    ];
```
Comment (Mic92) [inline nixos/modules/system/boot/systemd.nix:645]: Nitpick. My brain and nix processes this code much faster (also less parentheses):

```suggestion
    boot.kernelParams = optional (cfg.machineId != null) "systemd.machine_id=${cfg.machineId}"
      ++ optional (!cfg.enableUnifiedCgroupHierarchy) "systemd.unified_cgroup_hierarchy=0";
```

**Review by flokli — CHANGES_REQUESTED** [review]
> > Having a static value part of your system config is only a big footgun, having two machines with the same UUID. I don't think we should introduce this as a config option.
> 
> I don't see how it's more of a footgun than the plethora of extent options for setting a MAC address (worse, those have the potential for breaking the entire LAN the devices sit on) but I won't insist either; close as “won't fix” if the maintainers do not wish to provide such an option.

The machine-id is even more considered state:

> The machine ID is usually generated from a random source during system installation or first boot and stays constant for all subsequent boots. 

Pease pass it in externally where you have to keep it static across reboots, but can't keep state that state on the machine (for example netboot scenarios by appending via kernel cmdline), use it as state wherever you can (impermanence setups), and only resort to setting the kernel cmdline via NixOS options where you can't do otherwise. I don't think there should be a NixOS config option for this.

**Comment by arianvp** [comment]
I'm also not a big fan of this change.

https://systemd.io/BUILDING_IMAGES/ also recommends people not setting machine-id when building images.

Also there are multiple ways for people to set this.  (kernel command line vs  environment.etc). I think people should just use environment.etc or the kernel command line if they want to mess with this.

**Comment by RaitoBezarius** [comment]
@nbraud I would recommend the way forward is to document it in the NixOS manual about which situations may require to use a persistent `machine-id` you pass on the command line and to provide a simple snippet to do so.
More involved situations should probably have an out-of-tree NixOS module to perform the regex testing if needed.

**Comment by RaitoBezarius** [comment]
> > document it in the NixOS manual about which situations may require to use a persistent `machine-id` you pass on the command line and to provide a simple snippet to do so.
> 
> @RaitoBezarius What part of the manual do you think this should go in?

This could go in https://nixos.org/manual/nixos/stable/#ch-running under its own section potentially.

**Pull request closed without merging**